Move from boa to rattler-build

peterfpeterson opened the pull request:
Move the packaging to use rattler-build and all the niceties that come with it. Part of the headline improvements is decreased build times, mostly due to dependency solving.

The post-packaging warnings for [overdepending are explained here](https://rattler.build/v0.35.0/build_options/#dynamic-linking-configuration).

Fixes #39174.

<!-- If the original issue was raised by a user they should be named here. Do not leak email addresses
**Report to:** [user name]
-->

### To test:
Try outputing into a separate directory to avoid a recursive copy
```sh
cd conda/recipe/
rattler-build build -r ./mantid-developer -m ./conda_build_config.yaml --output-dir=/home/me/tmp/rattler
rattler-build build -r ./mantid -m ./conda_build_config.yaml --output-dir=/home/me/tmp/rattler
rattler-build build -r ./mantidqt -m ./conda_build_config.yaml --output-dir=/home/me/tmp/rattler
rattler-build build -r ./mantiddocs -m ./conda_build_config.yaml --output-dir=/home/me/tmp/rattler
rattler-build build -r ./mantidworkbench -m ./conda_build_config.yaml --output-dir=/home/me/tmp/rattler
```
This will generate packages with version number "0.0.0". To set the version number in the package, one must inject it through an environment variable (like the jenkins buildscript) `MANTID_VERSION=$(versioningit ../../)`

Then enter `/home/me/tmp/ratter/` and install the package(s). Note that `mantidworkbench` does not require `mantiddocs`, but if `mantiddocs` is installed the version must match.
```sh
mamba install -c conda-forge conda-build conda-index
python -m conda_index .
mamba env create -n test -c .  mantidworkbench mantiddocs
mamba activate test
mantidworkbench
```
Which will bring up mantid workbench.

<!-- delete this if you added release notes
*This does not require release notes* because **fill in an explanation of why**
If you add release notes please save them as a separate file using the Issue or PR number as the file name. Check the file is located in the correct directory for your note(s).
-->

<!-- Ensure the base of this PR is correct (e.g. release-next or main)
Finally, don't forget to add the appropriate labels, milestones, etc.!  -->

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

peterfpeterson committed:
Convert mantid framework recipe to rattler-build

peterfpeterson committed:
Convert mantid developer recipe to rattler-build

peterfpeterson committed:
Add rattler-build

peterfpeterson committed:
Change from boa to rattler-build

peterfpeterson committed:
Add rattler-build to the base environment

peterfpeterson committed:
Remove debug option

peterfpeterson committed:
Change the other recipes to build with rattler

peterfpeterson committed:
Reduce number of ifs

peterfpeterson committed:
Inject mantid version number

thomashampson commented on conda/recipes/conda_build_config.yaml:
```diff
@@ -60,26 +60,18 @@ muparser:
 
 ninja:
   - '>=1.10.2'
 
```
I think this was removed in error
```suggestion
# Fix the version to avoid updates being pulled in automatically, which might change the Reflectometry ORSO file content or layout and cause tests to fail.
orsopy:
  - 1.2.1

```

thomashampson commented on conda/recipes/mantid-developer/recipe.yaml:
```diff
@@ -0,0 +1,120 @@
+      then:
+        - clang_osx-64 ${{ c_compiler_version }}
+        - clangxx_osx-64 ${{ cxx_compiler_version }}
+        - clang_osx-arm64 ${{ c_compiler_version }}
+        - clangxx_osx-arm64 ${{ cxx_compiler_version }}
```
These need to be in a separate if block, otherwise it's using the system compiler, which is casuing the macOS failure
```suggestion
    - if: osx and arm64
      then:
        - clang_osx-arm64 ${{ c_compiler_version }}
        - clangxx_osx-arm64 ${{ cxx_compiler_version }}
```

peterfpeterson committed:
Restore orsopy pin

peterfpeterson committed:
Separate arm and intel osx

peterfpeterson committed:
Manually verified against meta.yaml in main

peterfpeterson committed:
Synchronize the opengl dependencies

peterfpeterson committed:
Add pip to the mantiqt build

peterfpeterson committed:
Move pip to the host section

peterfpeterson committed:
Install offline docs

peterfpeterson committed:
Add mantid and mantidqt to host requirements

peterfpeterson committed:
Added pip to host dependencies

peterfpeterson committed:
Require mantid framework for runtime

peterfpeterson committed:
Fix path for mantid.png

This was verified against the boa generated package mantiddocs-6.13.20250729.1217-hb55aea5_0.tar.bz2
It is unclear why boa didn't catch the error, but rattler did

peterfpeterson committed:
Add conda-index to make installing locally built packages easier. conda-verify does not currently support new style conda packages.

peterfpeterson committed:
Add occt back and comment on the build variant

peterfpeterson committed:
Set novtk in run environment

peterfpeterson committed:
Follow new filename convention

peterfpeterson committed:
Update developer docs to reflect new tooling

peterfpeterson committed:
Move osx-64 information to rattler-build argument

peterfpeterson committed:
Clarify conda activate failures

thomashampson commented on conda/recipes/mantid/recipe.yaml:
```diff
@@ -0,0 +1,120 @@
+      then: lib3mf
+    - muparser ${{ muparser }}
+    - ${{ pin_compatible("numpy", upper_bound="2.2") }}
+    # this should be set in the hosts as well, but pin_compatible doesn't currently support that
+    # https://github.com/prefix-dev/rattler-build/issues/1820
+    - ${{ pin_compatible("occt", build="*novtk*") }}
```
```suggestion
    - occt ${{ occt }}
```

thomashampson reviewed (dismissed):
It turns out that we can pin to the novtk build variant using the syntax:
```
occt:
  - '* novtk*'
```

thomashampson committed:
Pin occt to the novtk variant in all environments

peterfpeterson committed:
Add more import tests to ensure mantid packages are required correctly

peterfpeterson committed:
Remove unused variable

peterfpeterson committed:
Remove previous buildtree

This should improve stability in the build runs. Since each run is
completely separate, nothing should be reused anyhow.

peterfpeterson committed:
rattler doesn't write logs so there is nothing to archive

peterfpeterson committed:
Change filename from bld.bat to build.bat

This is part of the instructions from https://rattler.build/latest/converting_from_conda_build/

peterfpeterson committed:
Do not override environment variables to see what happens

rboston628 commented:
I was able to build mantid-developer, mantid, and mantidqt on my Intel MacBook (didn't try others, but sure they would work as well).  Mine is an 8-core Intel Core i9, and it runs macOS Version 12.4 (Monterey).  I built from within the `mantid-developer` mamba environment after adding `rattler-build` to it and following the above "To Test" instructions.

peterfpeterson committed:
Set a minimum pin for rattler with hopes that a new version will help

peterfpeterson committed:
Fix linker warning on osx-64

peterfpeterson committed:
Overide deployment target

peterfpeterson committed:
Override mac compatibility in the wrong place - this should break linux

peterfpeterson committed:
Add a test for getting workbench version

peterfpeterson committed:
Check that building on intel mac before setting variable

peterfpeterson committed:
Set sip version to match other platforms

peterfpeterson committed:
Fix syntax for seekpath build number

peterfpeterson committed:
Change how osx cmake variable gets set

peterfpeterson committed:
Remove support for packaging on osx-64 / intel mac

peterfpeterson committed:
Clean up boost versioning

peterfpeterson committed:
Fixup mistakes from rebasing

peterfpeterson committed:
re-sync with what is in main

peterfpeterson committed:
Sync with main ... again

peterfpeterson committed:
Convert the conda recipes to rattler recipes again

peterfpeterson committed:
Add pip back into mantidqt

peterfpeterson committed:
Change syntax of pin for windoze dependency

peterfpeterson committed:
Pin to rattler-build 0.47

peterfpeterson committed:
Fix call to get environment variable

peterfpeterson committed:
Add pip back to mantidworkbench

peterfpeterson committed:
Get mantiddocs target building again

MialLewis committed:
add log output and remove error handling

MialLewis committed:
output prefixes to file

MialLewis committed:
move archiving out of post stage

MialLewis committed:
move env_logs outside of conda-bld

peterfpeterson committed:
Add cmake args from rattler for windows

peterfpeterson committed:
Increase max path length from 250 to 300

MialLewis committed:
switch back to using msbuild as windows generator

MialLewis committed:
move mantid_version string generation

MialLewis committed:
move archive logs back to post

MialLewis committed:
commit submitted with no comment...

MialLewis committed:
Revert "move env_logs outside of conda-bld"

This reverts commit 1bd9037f1e751e9edee70ea2e28624b9ad4a79cb.

uncomment archive_env_logs

remove mistake

MialLewis committed:
remove toolset from cmakepresents in an attempt to fix windows PR job

MialLewis committed:
readd package-conda change

MialLewis committed:
use rattler-index

MialLewis committed:
correct base env logging

remove debug

MialLewis committed:
Merge remote-tracking branch 'origin/main' into rattler_build

MialLewis committed:
add missing if closure

MialLewis committed:
correct pin syntax

MialLewis committed:
add build options required to use VS generator

MialLewis committed:
correct archive scripts

adriazalvarez committed:
Change some restrictions on input validation of SANS Polarization Corrections Algorithm (#40100)

### Description of work
This PR fixes #40099 . More details can be found in issue. 

Additionally, as this algorithm is being tested, I have tried to
increase the flexibility for diagnostics in this algorithm, adding a
check for avoiding deleting partial workspaces if the algorithm fails,
as well as allowing to set initial values for some of the fitting
properties.

<!-- Please provide an outline and reasoning for the work.
If there is no linked issue provide context.
-->

Closes #40099. <!-- One line per closed issue. -->
- Test are passing. 
- Contact me for test data.

<!-- If issue raised by user. Do not leak email addresses.
**Report to:** [user name]
-->

### To test:

<!-- Include sufficient instructions for someone unfamiliar with the
application to test.
Ok to refer back to instructions in the issue.
-->

Co-authored-by: Jonathan Haigh <35813666+jhaigh0@users.noreply.github.com>

MialLewis committed:
add build options to run-tests to use VS generator

MialLewis committed:
tidy up mantid-developer recipe

jhaigh0 committed:
Framework release notes for 6.14 (#40093)

### Description of work

The note reading 
```
- :ref:`SNSPowderReduction<algm-SNSPowderReduction>` has been updated to more correctly apply absorption corrections.
```
has been moved to the diffraction notes
https://github.com/mantidproject/mantid/pull/40091

### To test:

- Spelling and grammar
- Check the file builds correctly and all links etc work. 

Co-authored-by: Caila Finn <47181718+cailafinn@users.noreply.github.com>

MialLewis committed:
tidy up remaining recipies

MialLewis committed:
readd toolset

jhaigh0 committed:
Indirect release notes 6.14 (#40094)

### Description of work

i've moved the one and only release note from inelastic to this page to
keep the Abins related notes together. (Have spoken to Adam and he's
happy with this).

### To test:

- Check spelling and grammar
- Build `docs-html` to check formatting

jhaigh0 committed:
LSS release notes 6.14 (#40096)

### Description of work

Combined reflectometry and sans since there was only one reflectomery
change.

### To test:

- Spelling and grammar check
- Build `docs-html` and check formatting

jhaigh0 committed:
Direct release notes 614 (#40092)

### Description of work

One note has moved to `framework.rst`, the rest of edits to
`framework.rst` will come in another pr.

### To test:

- Spelling and grammar checks
- Build `docs-html` and check everything looks correct

jhaigh0 committed:
Raise open file descriptors limit to avoid problem with pystack (#40108)

### Description of work

The recent problems with running `pystack` in the error reporter have
been traced back to the fact that `pystack` was trying to open more file
descriptors than the limit set on most linux systems. A pr into
`pystack` has been made to improve their error messages
(https://github.com/bloomberg/pystack/pull/260).

When this is a problem, you get a lot of this kind of output
```
ERROR(process_core): Cannot open ELF file /home/bya67386/mambaforge/envs/test_htlp_links/lib/python3.11/site-packages/mantid/kernel/../../../.././libMantidJson.so
ERROR(process_core): Cannot open ELF file /home/bya67386/mambaforge/envs/test_htlp_links/lib/python3.11/site-packages/mantid/kernel/../../../.././libMantidJson.so
ERROR(process_core): Cannot open ELF file /home/bya67386/mambaforge/envs/test_htlp_links/lib/python3.11/site-packages/mantid/kernel/../../../.././libMantidJson.so
ERROR(process_core): Cannot open ELF file /home/bya67386/mambaforge/envs/test_htlp_links/lib/python3.11/site-packages/mantid/kernel/../../../.././libMantidJson.so
ERROR(process_core): Cannot open ELF file /home/bya67386/mambaforge/envs/test_htlp_links/lib/python3.11/site-packages/mantid/kernel/../../../../libMantidTypes.so
ERROR(process_core): Cannot open ELF file /home/bya67386/mambaforge/envs/test_htlp_links/lib/python3.11/site-packages/mantid/kernel/../../../../libMantidTypes.so
ERROR(process_core): Cannot open ELF file /home/bya67386/mambaforge/envs/test_htlp_links/lib/python3.11/site-packages/mantid/kernel/../../../../libMantidTypes.so
ERROR(process_core): Cannot open ELF file /home/bya67386/mambaforge/envs/test_htlp_links/lib/python3.11/site-packages/mantid/kernel/../../../../libMantidTypes.so
ERROR(process_core): Cannot open ELF file /home/bya67386/mambaforge/envs/test_htlp_links/lib/python3.11/site-packages/mantid/kernel/../../../../libMantidPythonInterfaceCore.so
ERROR(process_core): Cannot open ELF file /home/bya67386/mambaforge/envs/test_htlp_links/lib/python3.11/site-packages/mantid/kernel/../../../../libMantidPythonInterfaceCore.so
ERROR(process_core): Cannot open ELF file /home/bya67386/mambaforge/envs/test_htlp_links/lib/python3.11/site-packages/mantid/kernel/../../../../libMantidPythonInterfaceCore.so
ERROR(process_core): Cannot open ELF file /home/bya67386/mambaforge/envs/test_htlp_links/lib/python3.11/site-packages/mantid/kernel/../../../../libMantidPythonInterfaceCore.
```

The issue started after the first new instrument view pr went in (this
commit
https://github.com/mantidproject/mantid/commit/8733e98a3c0b36e46ccd99b612ecca1f00546b7f#diff-4a6d6571d2c59705d8a673667672de2d7dcc3c756c61a3f0195acdefac52fe34).
We're not that sure why, but it could be the addition of new libraries
such as `pyvista`.

This pr raises the limit (equivalent to `ulimit -n <value>`) for the
outer mantid process, which may eventually launch the error reporter.

### To test:

Package to test is at `mamba install
jhaigh0/label/raise_open_files_limit::mantidworkbench`

It would be good to test on idaaas where the core dump location is
already set up.
Run Segfault and check that the pystack output has been captured in the
stacktrace field.

jhaigh0 committed:
Mantidworkbench release notes 6.14 (#40105)

### Description of work
A couple of notes about the new python property have gone to the
framework notes

### To test:

- Spelling and grammar
- Builds with good formatting

Co-authored-by: thomashampson <thomas.hampson@stfc.ac.uk>

jclarkeSTFC committed:
Replaced uses of getNumberBins() with blocksize() (#40029)

`MatrixWorkspace::getNumberBins` has been deprecated, replaced uses with
`MatrixWorkspace::blocksize`

See
https://github.com/mantidproject/mantid/issues/39987#issuecomment-3347444053

Found during manual testing for 6.14

### To test:

See
https://github.com/mantidproject/mantid/issues/39987#issuecomment-3347444053
for testing instructions.

jclarkeSTFC commented on buildconfig/Jenkins/Conda/package-conda:
```diff
@@ -138,22 +141,19 @@ mamba run -n base mamba list
 fi
 if [[ $ENABLE_BUILD_WORKBENCH == true ]]; then
-    if [[ $BUILD_WORKBENCH_WITH_DOCS == false ]]; then
-	export INCLUDE_MANTIDDOCS=False
```
Is `INCLUDE_MANTIDDOCS` not required any more?

jclarkeSTFC commented on conda/recipes/mantid-developer/recipe.yaml:
```diff
@@ -0,0 +1,119 @@
+    - quasielasticbayes ${{ quasielasticbayes }}
+    - quickbayes ${{ quickbayes }}
+    # something is wrong with versioningit and rattler-build 0.48
+    - rattler-build =0.47
```
Do we need `rattler-build` in `mantid-developer`? I think that building the packages is best done is a separate environment without `mantid-developer`

andy-bridger committed:
Texture analysis Concept Document (#39643)

### Description of work

#### Summary of work
Contains some background on texture analysis and pole figures as well as
some scripts for creating pole figures once #39581 is merged

#### Purpose of work
to inform

### To test:

Once #39581 is merged, the scripts provided could be tested (I can
provide any data files needed).

Otherwise read through the page and check for grammatical issues or
anything that doesn't make sense.

*This does not require release notes* because **no change to mantid
functionality**

jclarkeSTFC commented on conda/recipes/mantidqt/recipe.yaml:
```diff
@@ -0,0 +1,83 @@
+      then:
+        - libopenblas ${{ libopenblas }}
+  run:
+    - mantid == ${{ version }}
```
If `mantid` is in `host`, can we get rid of it in `run`?

peterfpeterson replied:
I believe mantid is in the host section so it can be used during the setup/configure stage of things. mantid in run is so users get it. Things in host don't automatically appear in run.

jclarkeSTFC commented on conda/recipes/mantidworkbench/recipe.yaml:
```diff
@@ -0,0 +1,87 @@
+    - setuptools
+    - mantidqt == ${{ version }}
+    - python ${{ python }}
+    - pip ${{ pip }}
```
`pip` is a new addition, why is it needed?

jclarkeSTFC reviewed (commented):
Builds `mantid-developer` in <1 second on my laptop, way faster than before. Windows standalone worked okay, tried opening a plot and running an algorithm. Jenkins build branch run looked way faster as well. I didn't see anything wrong, but did have a few questions.

MialLewis committed:
Fix linthreshz key error when changing 3D plot background color (#40040)

### Description of work

This PR:
- Fixes an error thrown when a change to the plot background color.
"linthreshz" has been added to axes props. This whilst symlog is
disabled for 3D plots and therefore not applicable to the Z axis, this
is still used to set the value of the hidden `linthresh` field in the
background.
- Moves the default value on linthresh to a variable

Closes #40041

<!-- If issue raised by user. Do not leak email addresses.
**Report to:** [user name]
-->

### To test:

- Follow the test instructions in the issue.

thomashampson reviewed (changes requested):
This will make a huge difference to all of our CI workflows, looking forward to seeing it go in. However there is a typo in the jenkinsfile that causes the standalone packaging scripts to install packages from the nightly label rather than the local channel. Please can you commit my suggestion and re-run the test build to confirm that it picks up the correct packages?

peterfpeterson committed:
Change file extension for packages

Co-authored-by: thomashampson <thomas.hampson@stfc.ac.uk>

peterfpeterson committed:
fixup! Merge remote-tracking branch 'upstream/main' into rattler_build

peterfpeterson committed:
Merge remote-tracking branch 'upstream/main' into rattler_build

jclarkeSTFC reviewed (dismissed):
Downloaded the Windows installer, version matched the conda packages in the artifacts.

thomashampson commented:
I will merge this early next week once the smoke testing is underway

thomashampson committed:
Merge branch 'main' into rattler_build

thomashampson committed:
Fix mslice pin for rattler build

thomashampson reviewed (approved):
This has been thoroughly tested already to my satisfaction. The test failures on Linux are unrelated. The most recent Windows failure is understood.
